pkg/scheduler/podtask/ranges.go

```diff
@@ -0,0 +1,140 @@
+package podtask
+
+import (
+	"sort"
+
+	"github.com/gogo/protobuf/proto"
+	mesos "github.com/mesos/mesos-go/mesosproto"
+)
+
+// Range is an all-inclusive range defined by a lower and upper bound.
+type Range [2]uint64
+
+// Ranges represents a list of Ranges.
+type Ranges []Range
+
+// NewPortRanges returns Ranges from the "ports" resource in the
+// given *mesos.Offer. If that resource isn't provided, nil will be returned.
+//
+// The returned Ranges are sorted and have all overlapping ranges merged from
+// left to right. e.g. [[0, 5], [4, 3], [10, 7]] -> [[0, 5], [7, 10]]
```

Comment:
The documentation on the `mesos.proto` definition says:
```
/**
 * Describes a resource on a machine. A resource can take on one of
 * three types: scalar (double), a list of finite and discrete ranges
 * (e.g., [1-10, 20-30]), or a set of items. A resource is described
 * using the standard protocol buffer "union" trick.
 *
 * TODO(benh): Add better support for "expected" resources (e.g.,
 * cpus, memory, disk, network).
 */
```

Unfortunately this doesn't really guarantee if the ranges are sorted or squashed correctly. The **discreteness** alluded to could have no gaps (e.g. `[[1, 10], [11,20]]`) which should be squashed (`[[1, 20]]`).

I wend down this road to for the sake of robustness. We should try to avoid relying on non-documented behaviour or implementation details.

Now, I totally agree this could be part of `mesos-go`! Good call. Though, I'm not sure if we should spend time on that right now. Perhaps we can keep it in our roadmap.